### PR TITLE
Turn validation problem from warning to error: no cacheability annotation on task type

### PR DESCRIPTION
**Pull request by lptr**
Turn this validation problem into an error: https://github.com/gradle/gradle/blob/f68a83ff2b3576fa3e1f847332f6134f01826620/subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/tasks/internal/ValidateAction.java#L149-L162

See https://github.com/gradle/gradle/issues/21648

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc [hunk 1]
```diff
@@ -52,6 +52,11 @@ Creating a TAR tree from a resource with no backing file is no longer supported.
 Instead, convert the resource to a file and use `project.tarTree()` on the file.
 For more information, see <<tar_tree_no_backing_file>>.
 
+==== Task types without explicit cacheability annotations
+
+All task types must now be annotated with either `@CacheableTask`, `@DisableCachingByDefault` or `@UntrackedTask`.
+Having neither annotation present results in an error.
```
Comment (nathan-contino) [inline subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc:58]:
```suggestion
Declaring a task type without one of these annotations produces an error.
```
Suggestion: reword for clarity

**Review by nathan-contino — APPROVED** [review]
docs LGTM, one suggestion

File: subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc (hunk 1 above)
Comment (lptr) [reply subprojects/docs/src/docs/userguide/migration/upgrading_version_7.adoc]: I've updated the doc a bit, PTAL.

**Review by nathan-contino — APPROVED** [review]
Docs LGTM

**Review by alllex — APPROVED** [review]
LGTM!

**Comment by lptr** [comment]
Superseded by https://github.com/gradle/gradle/pull/21793

**Pull request closed without merging**